invenio_rdm_migrator/state.py

```diff
@@ -0,0 +1,248 @@
+
+        # move current db state to backup
+        backup_file.unlink(missing_ok=True)
+
+        end = ts(iso=False)
+        logger.info(f"Finished dumping state {ts()}")
+        logger.info(f"State dumping took {end-start} seconds.")
+
+    def get(self, table_name, key_attr, key_value):
+        """Query a table by key."""
+        table = self.metadata.tables[table_name]
+
+        return self.mem_eng.execute(
+            select(table).where(getattr(table.columns, key_attr) == key_value)
+        ).one_or_none()
+
+    def all(self, table_name):
+        """Get all the data from the state."""
+        table = self.metadata.tables[table_name]
+
+        return self.mem_eng.execute(select(table)).all()
+
+    def add(self, table_name, data):
+        """Add key,data pair to the state."""
+        table = self.metadata.tables[table_name]
+        self._validate(table, data)
+        res = self.mem_eng.execute(insert(table).values(**data))
+        assert res.rowcount == 1
+
+    def update(self, table_name, key_attr, key_value, data):
+        """Update an entry."""
+        table = self.metadata.tables[table_name]
+
+        self._validate(table, data)
+        res = self.mem_eng.execute(
+            update(table)
+            .where(getattr(table.columns, key_attr) == key_value)
+            .values(**data)
+        )
+        assert res.rowcount == 1  # the update succeeded
+
+    def _copy_db(cls, src, dest):
+        """Copy source database into destination."""
+        # sqlalchemy abstraction does not have a backup we need a raw connection
+        src_conn = src.raw_connection()
+        dst_conn = dest.raw_connection()
+
+        src_conn.backup(dst_conn.connection)
+
+        src_conn.close()
+        dst_conn.close()
+
+    def _validate(self, table, data):
+        """Validate data for a specific table."""
+        validator = self.validators.get(table.name)
+        if validator:
+            validator.validate(data)
+
+    @classmethod
+    def _initialize_db(cls, metadata):
+        """Initialize state tables and bind to metadata object."""
+        # this registration method is not ideal. It implies a tight coupling between
+        # state and which streams can be run, and there is no independence (i.e. running one
+        # stream means having all state tables defined rather than only those needed).
+        # The tables should be defined per stream (e.g. streams/records/state.py:ParentTable)
+        # and then be injected into the state.
+
+        Table(
+            "parents",
+            metadata,
+            Column("recid", String, primary_key=True),
+            Column("id", UUIDType, nullable=False, unique=True),
+            Column("latest_id", UUIDType),
+            # could use default=0, but it could create uncontrolled scenarios
+            Column("latest_index", Integer),
+            Column("next_draft_id", UUIDType),
+            sqlite_autoincrement=False,
+        )
```

Comment:
I used core for the connections due to the fact that I had to play with the memory/disk loading/dumping. Then it seemed there was not much gained from using the ORM since our queries/interactions were quite simple. But I can modify that, WDYT?

Reply:
It looks pretty clean tbh, and as you said it's simple, so was just wondering if there was a drawback or blocker for the ORM. Fine with me to use Core